File: utils/calculations.py
"""
Anthropometric and nutritional calculations.
All weight in kg, height in cm, circumferences in cm, skinfolds in mm.
"""
import math
from typing import Optional, Tuple


ACTIVITY_FACTORS = {
    "Sedentario (sin ejercicio)": 1.2,
    "Ligeramente activo (1-3 dias/sem)": 1.375,
    "Moderadamente activo (3-5 dias/sem)": 1.55,
    "Muy activo (6-7 dias/sem)": 1.725,
    "Extremadamente activo (atleta)": 1.9,
}

# Durnin & Womersley (1974) constants: (min_age, max_age, C, M)
# Body density = C - M × log10(Σ4 skinfolds)
# Σ4 = biceps + triceps + subscapular + cresta iliaca
_DW_WOMEN = [
    (17, 19, 1.1549, 0.0678),
    (20, 29, 1.1599, 0.0717),
    (30, 39, 1.1423, 0.0632),
    (40, 49, 1.1333, 0.0612),
    (50, 999, 1.1339, 0.0645),
]
_DW_MEN = [
    (17, 19, 1.1620, 0.0630),
    (20, 29, 1.1631, 0.0632),
    (30, 39, 1.1422, 0.0544),
    (40, 49, 1.1620, 0.0700),
    (50, 999, 1.1715, 0.0779),
]


def sum_6_skinfolds(triceps, subscapular, supraspinal,
                    abdominal, medial_thigh, max_calf) -> Optional[float]:
    """Sum of 6 ISAK skinfolds (marked with *). All in mm."""
    vals = [triceps, subscapular, supraspinal, abdominal, medial_thigh, max_calf]
    if any(v is None for v in vals):
        return None
    return round(sum(vals), 1)


def body_fat_durnin_womersley(biceps_mm: float, triceps_mm: float,
                               subscapular_mm: float, iliac_crest_mm: float,
                               age: int, sex: str) -> Tuple[Optional[float], Optional[float]]:
    """
    Durnin & Womersley (1974) — ISAK Level 1.
    Σ4 = bíceps + tríceps + subescapular + cresta iliaca.
    Returns (body_density, fat_mass_pct) via Siri equation.
    """
    try:
        sigma4 = biceps_mm + triceps_mm + subscapular_mm + iliac_crest_mm
        if sigma4 <= 0:
            return None, None
        log_s = math.log10(sigma4)

        table = _DW_MEN if sex.lower() in ("masculino", "m", "hombre") else _DW_WOMEN
        C, M = table[-1][2], table[-1][3]   # default: oldest range
        for min_a, max_a, c, m in table:
            if min_a <= age <= max_a:
                C, M = c, m
                break

        density = C - M * log_s
        fat_pct = (4.95 / density - 4.5) * 100
        return round(density, 4), round(fat_pct, 1)
    except (ValueError, ZeroDivisionError):
        return None, None


def fat_mass_kg(weight_kg: float, fat_pct: float) -> float:
    return round(weight_kg * fat_pct / 100, 2)


def lean_mass_kg(weight_kg: float, fat_kg: float) -> float:
    return round(weight_kg - fat_kg, 2)


# ── Classic formulas (kept for full report compatibility) ────────────────────

def bmi(weight_kg: float, height_cm: float) -> float:
    if height_cm <= 0:
        return 0.0
    return round(weight_kg / (height_cm / 100) ** 2, 2)


def bmi_category(bmi_val: float) -> str:
    if bmi_val < 18.5:
        return "Bajo peso"
    elif bmi_val < 25.0:
        return "Normal"
    elif bmi_val < 30.0:
        return "Sobrepeso"
    elif bmi_val < 35.0:
        return "Obesidad grado I"
    elif bmi_val < 40.0:
        return "Obesidad grado II"
    else:
        return "Obesidad grado III"


def ideal_weight_lorentz(height_cm: float, sex: str) -> float:
    if height_cm <= 0:
        return 0.0
    if sex.lower() in ("masculino", "m", "hombre"):
        return round(height_cm - 100 - (height_cm - 150) / 4, 1)
    return round(height_cm - 100 - (height_cm - 150) / 2, 1)


def bmr_mifflin(weight_kg: float, height_cm: float, age: int, sex: str) -> float:
    base = 10 * weight_kg + 6.25 * height_cm - 5 * age
    if sex.lower() in ("masculino", "m", "hombre"):
        return round(base + 5, 1)
    return round(base - 161, 1)


def tdee(bmr_val: float, activity_level: str) -> float:
    return round(bmr_val * ACTIVITY_FACTORS.get(activity_level, 1.2), 1)


def macro_distribution(total_calories: float,
                        protein_pct: float = 30,
                        carbs_pct: float = 45,
                        fat_pct: float = 25) -> dict:
    return {
        "protein_g": round(total_calories * protein_pct / 100 / 4, 1),
        "carbs_g":   round(total_calories * carbs_pct   / 100 / 4, 1),
        "fat_g":     round(total_calories * fat_pct     / 100 / 9, 1),
    }


def waist_hip_ratio(waist_cm: float, hip_cm: float) -> float:
    return round(waist_cm / hip_cm, 3) if hip_cm > 0 else 0.0


def whr_risk(ratio: float, sex: str) -> str:
    is_male = sex.lower() in ("masculino", "m", "hombre")
    if is_male:
        return "Riesgo bajo" if ratio < 0.90 else ("Riesgo moderado" if ratio <= 0.99 else "Riesgo alto")
    return "Riesgo bajo" if ratio < 0.80 else ("Riesgo moderado" if ratio <= 0.85 else "Riesgo alto")


def body_fat_navy(sex: str, height_cm: float,
                  waist_cm: float, neck_cm: float,
                  hip_cm: float = 0.0) -> Optional[float]:
    try:
        if sex.lower() in ("masculino", "m", "hombre"):
            if waist_cm <= neck_cm:
                return None
            val = 495 / (1.0324 - 0.19077 * math.log10(waist_cm - neck_cm)
                         + 0.15456 * math.log10(height_cm)) - 450
        else:
            if (waist_cm + hip_cm) <= neck_cm:
                return None
            val = 495 / (1.29579 - 0.35004 * math.log10(waist_cm + hip_cm - neck_cm)
                         + 0.22100 * math.log10(height_cm)) - 450
        return round(val, 1)
    except (ValueError, ZeroDivisionError):
        return None


def body_fat_category(pct: float, sex: str, age: int) -> str:
    is_male = sex.lower() in ("masculino", "m", "hombre")
    if is_male:
        if pct < 6:   return "Esencial"
        if pct < 14:  return "Atleta"
        if pct < 18:  return "En forma"
        if pct < 25:  return "Aceptable"
        return "Obesidad"
    else:
        if pct < 14:  return "Esencial"
        if pct < 21:  return "Atleta"
        if pct < 25:  return "En forma"
        if pct < 32:  return "Aceptable"
        return "Obesidad"


# ── ISAK Level 2 formulas ────────────────────────────────────────────────────

def somatotype_heath_carter(
    height_cm: float, weight_kg: float,
    triceps_mm: float, subscapular_mm: float, supraspinal_mm: float,
    humerus_width_cm: float, femur_width_cm: float,
    arm_contracted_cm: float, calf_cm: float, max_calf_mm: float,
) -> tuple:
    """
    Heath & Carter (1990) somatotype.
    Returns (endomorphy, mesomorphy, ectomorphy) or (None, None, None).
    Requires diameters in cm, skinfolds in mm, perimeters in cm.
    """
    vals = [height_cm, weight_kg, triceps_mm, subscapular_mm, supraspinal_mm,
            humerus_width_cm, femur_width_cm, arm_contracted_cm, calf_cm, max_calf_mm]
    if any(v is None for v in vals):
        return None, None, None
    try:
        # Endomorphy — Σ3 skinfolds corrected to 170.18 cm height
        sigma3 = triceps_mm + subscapular_mm + supraspinal_mm
        sigma3_corr = sigma3 * (170.18 / height_cm)
        endo = (-0.7182 + 0.1451 * sigma3_corr
                - 0.00068 * sigma3_corr ** 2
                + 0.0000014 * sigma3_corr ** 3)

        # Mesomorphy — corrected arm and calf perimeters
        arm_corr = arm_contracted_cm - triceps_mm / 10
        calf_corr = calf_cm - max_calf_mm / 10
        meso = (0.858 * humerus_width_cm + 0.601 * femur_width_cm
                + 0.188 * arm_corr + 0.161 * calf_corr
                - height_cm * 0.131 + 4.5)

        # Ectomorphy via Height-Weight Ratio
        hwr = height_cm / (weight_kg ** (1 / 3))
        if hwr > 40.75:
            ecto = 0.732 * hwr - 28.58
        elif hwr > 38.25:
            ecto = 0.463 * hwr - 17.63
        else:
            ecto = 0.1

        return round(max(0.1, endo), 2), round(max(0.1, meso), 2), round(max(0.1, ecto), 2)
    except (TypeError, ValueError, ZeroDivisionError):
        return None, None, None


def waist_height_ratio(waist_cm: float, height_cm: float) -> Optional[float]:
    """Índice cintura/talla. Saludable < 0.5."""
    try:
        if not waist_cm or not height_cm:
            return None
        return round(waist_cm / height_cm, 3)
    except (TypeError, ZeroDivisionError):
        return None


# ── Puntos de corte / clasificación ─────────────────────────────────────────

LEVEL_COLOR = {
    "excellent": "#16a34a",
    "good":      "#16a34a",
    "average":   "#ca8a04",
    "high":      "#ea580c",
    "very_high": "#dc2626",
    "low":       "#2563eb",
    "moderate":  "#ca8a04",
    "risk":      "#dc2626",
}

LEVEL_EMOJI = {
    "excellent": "🟢",
    "good":      "🟢",
    "average":   "🟡",
    "high":      "🟠",
    "very_high": "🔴",
    "low":       "🔵",
    "moderate":  "🟡",
    "risk":      "🔴",
}

# Fat% tables: (min_age, max_age, [excellent_max, good_max, average_max, high_max])
_FAT_WOMEN_TABLE = [
    (20, 29, [15, 19, 28, 31]),
    (30, 39, [16, 20, 29, 32]),
    (40, 49, [17, 21, 30, 33]),
    (50, 59, [18, 22, 31, 34]),
    (60, 999, [19, 23, 32, 35]),
]
_FAT_MEN_TABLE = [
    (20, 29, [11, 13, 20, 23]),
    (30, 39, [12, 15, 21, 24]),
    (40, 49, [14, 17, 23, 26]),
    (50, 59, [15, 18, 24, 27]),
    (60, 999, [16, 19, 25, 28]),
]


def classify_fat_pct(pct: float, sex: str, age: int) -> tuple:
    """Returns (category_text, level_key) for % body fat."""
    is_male = sex.lower() in ("masculino", "m", "hombre")
    table = _FAT_MEN_TABLE if is_male else _FAT_WOMEN_TABLE
    t = table[-1][2]
    for min_a, max_a, thresh in table:
        if min_a <= age <= max_a:
            t = thresh
            break
    if pct <= t[0]: return "Excelente",    "excellent"
    if pct <= t[1]: return "Bueno",        "good"
    if pct <= t[2]: return "Promedio",     "average"
    if pct <= t[3]: return "Elevado",      "high"
    return "Muy elevado", "very_high"


def classify_bmi(bmi_val: float) -> tuple:
    """Returns (category_text, level_key) for BMI."""
    if bmi_val < 18.5: return "Bajo peso",   "low"
    if bmi_val < 25.0: return "Normal",       "excellent"
    if bmi_val < 30.0: return "Sobrepeso",    "average"
    if bmi_val < 35.0: return "Obesidad I",   "high"
    if bmi_val < 40.0: return "Obesidad II",  "very_high"
    return "Obesidad III", "very_high"


def classify_whr(ratio: float, sex: str) -> tuple:
    """Returns (category_text, level_key) for waist-hip ratio."""
    is_male = sex.lower() in ("masculino", "m", "hombre")
    if is_male:
        if ratio < 0.95:  return "Riesgo bajo",     "excellent"
        if ratio <= 1.00: return "Riesgo moderado",  "moderate"
        return "Riesgo alto", "risk"
    else:
        if ratio < 0.80:  return "Riesgo bajo",     "excellent"
        if ratio <= 0.85: return "Riesgo moderado",  "moderate"
        return "Riesgo alto", "risk"


def classify_whtr(ratio: float) -> tuple:
    """Returns (category_text, level_key) for waist-height ratio."""
    if ratio < 0.50: return "Saludable",           "excellent"
    if ratio < 0.60: return "Riesgo incrementado",  "average"
    return "Riesgo alto", "very_high"


def arm_muscle_area(arm_relaxed_cm: float, triceps_mm: float) -> Optional[float]:
    """
    Área Muscular del Brazo (AMB) en cm².
    AMB = (C - π × T/10)² / (4π)  donde C = perím. brazo relajado (cm), T = tríceps (mm).
    """
    try:
        if arm_relaxed_cm is None or triceps_mm is None:
            return None
        T_cm = triceps_mm / 10
        muscle_perim = arm_relaxed_cm - math.pi * T_cm
        if muscle_perim <= 0:
            return None
        return round(muscle_perim ** 2 / (4 * math.pi), 2)
    except (TypeError, ValueError):
        return None

File: frontend/src/utils/calculations.ts
/**
 * Anthropometric calculations — TypeScript port of utils/calculations.py
 * All skinfolds in mm, weight in kg, height/perimeters/diameters in cm.
 */

// ── Durnin & Womersley (1974) tables ─────────────────────────────────────────
// Body density = C - M × log10(Σ4 skinfolds)
// Σ4 = biceps + triceps + subscapular + iliac_crest
const DW_WOMEN: [number, number, number, number][] = [
  [17, 19, 1.1549, 0.0678],
  [20, 29, 1.1599, 0.0717],
  [30, 39, 1.1423, 0.0632],
  [40, 49, 1.1333, 0.0612],
  [50, 999, 1.1339, 0.0645],
]

const DW_MEN: [number, number, number, number][] = [
  [17, 19, 1.162, 0.063],
  [20, 29, 1.1631, 0.0632],
  [30, 39, 1.1422, 0.0544],
  [40, 49, 1.162, 0.07],
  [50, 999, 1.1715, 0.0779],
]

// ── Types ─────────────────────────────────────────────────────────────────────

export interface IsAk1Result {
  sigma4: number
  body_density: number
  fat_mass_pct: number
  fat_mass_kg: number
  lean_mass_kg: number
  sum_6_skinfolds: number | null
}

export interface SomatotypeResult {
  endo: number
  meso: number
  ecto: number
}

// ── Helpers ───────────────────────────────────────────────────────────────────

function isMale(sex: string): boolean {
  const s = sex.toLowerCase()
  return s.includes('masculino') || s === 'm' || s === 'hombre'
}

function round(value: number, decimals: number): number {
  const f = 10 ** decimals
  return Math.round(value * f) / f
}

function validNum(v: number | null | undefined): boolean {
  return v != null && !isNaN(v) && isFinite(v)
}

// ── ISAK 1 — Durnin & Womersley (1974) ───────────────────────────────────────

/**
 * Calculates body density and fat % using Durnin & Womersley (1974).
 * Returns null if any required value is missing or invalid.
 */
export function calcIsAk1(
  biceps_mm: number | null,
  triceps_mm: number | null,
  subscapular_mm: number | null,
  iliac_crest_mm: number | null,
  weight_kg: number | null,
  age: number | null,
  sex: string,
  // Optional for sum_6 calculation
  supraspinal_mm?: number | null,
  abdominal_mm?: number | null,
  medial_thigh_mm?: number | null,
  max_calf_mm?: number | null,
): IsAk1Result | null {
  if (
    !validNum(biceps_mm) || !validNum(triceps_mm) ||
    !validNum(subscapular_mm) || !validNum(iliac_crest_mm) ||
    !validNum(weight_kg) || !validNum(age) ||
    biceps_mm! <= 0 || triceps_mm! <= 0 ||
    subscapular_mm! <= 0 || iliac_crest_mm! <= 0 ||
    weight_kg! <= 0 || age! <= 0
  ) return null

  const sigma4 = biceps_mm! + triceps_mm! + subscapular_mm! + iliac_crest_mm!
  if (sigma4 <= 0) return null

  const logS = Math.log10(sigma4)
  const table = isMale(sex) ? DW_MEN : DW_WOMEN

  let C = table[table.length - 1][2]
  let M = table[table.length - 1][3]
  for (const [minA, maxA, c, m] of table) {
    if (age! >= minA && age! <= maxA) { C = c; M = m; break }
  }

  const body_density = C - M * logS
  const fat_pct = (4.95 / body_density - 4.5) * 100
  const fat_kg = (weight_kg! * fat_pct) / 100
  const lean_kg = weight_kg! - fat_kg

  // Σ6 = triceps + subscapular + supraspinal + abdominal + medial_thigh + max_calf
  let sum6: number | null = null
  if (
    validNum(triceps_mm) && validNum(subscapular_mm) &&
    validNum(supraspinal_mm) && validNum(abdominal_mm) &&
    validNum(medial_thigh_mm) && validNum(max_calf_mm)
  ) {
    sum6 = round(
      triceps_mm! + subscapular_mm! + supraspinal_mm! +
      abdominal_mm! + medial_thigh_mm! + max_calf_mm!,
      1,
    )
  }

  return {
    sigma4: round(sigma4, 1),
    body_density: round(body_density, 4),
    fat_mass_pct: round(fat_pct, 1),
    fat_mass_kg: round(fat_kg, 2),
    lean_mass_kg: round(lean_kg, 2),
    sum_6_skinfolds: sum6,
  }
}

// ── ISAK 2 — Heath & Carter (1990) Somatotype ────────────────────────────────

/**
 * Heath & Carter (1990) somatotype.
 * Returns null if any required value is missing.
 */
export function calcSomatotype(
  height_cm: number | null,
  weight_kg: number | null,
  triceps_mm: number | null,
  subscapular_mm: number | null,
  supraspinal_mm: number | null,
  humerus_width_cm: number | null,
  femur_width_cm: number | null,
  arm_contracted_cm: number | null,
  calf_cm: number | null,
  max_calf_mm: number | null,
): SomatotypeResult | null {
  const vals = [
    height_cm, weight_kg, triceps_mm, subscapular_mm, supraspinal_mm,
    humerus_width_cm, femur_width_cm, arm_contracted_cm, calf_cm, max_calf_mm,
  ]
  if (vals.some((v) => !validNum(v) || v! <= 0)) return null

  try {
    // Endomorphy — Σ3 skinfolds corrected to 170.18 cm height
    const sigma3 = triceps_mm! + subscapular_mm! + supraspinal_mm!
    const sigma3_corr = sigma3 * (170.18 / height_cm!)
    const endo =
      -0.7182 +
      0.1451 * sigma3_corr -
      0.00068 * sigma3_corr ** 2 +
      0.0000014 * sigma3_corr ** 3

    // Mesomorphy — corrected arm and calf perimeters
    const arm_corr = arm_contracted_cm! - triceps_mm! / 10
    const calf_corr = calf_cm! - max_calf_mm! / 10
    const meso =
      0.858 * humerus_width_cm! +
      0.601 * femur_width_cm! +
      0.188 * arm_corr +
      0.161 * calf_corr -
      height_cm! * 0.131 +
      4.5

    // Ectomorphy via Height-Weight Ratio
    const hwr = height_cm! / Math.cbrt(weight_kg!)
    let ecto: number
    if (hwr > 40.75) ecto = 0.732 * hwr - 28.58
    else if (hwr > 38.25) ecto = 0.463 * hwr - 17.63
    else ecto = 0.1

    return {
      endo: round(Math.max(0.1, endo), 2),
      meso: round(Math.max(0.1, meso), 2),
      ecto: round(Math.max(0.1, ecto), 2),
    }
  } catch {
    return null
  }
}

// ── Área Muscular del Brazo ───────────────────────────────────────────────────

/**
 * AMB = (C - π × T/10)² / (4π)
 * C = perím. brazo relajado (cm), T = tríceps (mm)
 */
export function calcArmMuscleArea(
  arm_relaxed_cm: number | null,
  triceps_mm: number | null,
): number | null {
  if (!validNum(arm_relaxed_cm) || !validNum(triceps_mm)) return null
  const musclePerim = arm_relaxed_cm! - Math.PI * (triceps_mm! / 10)
  if (musclePerim <= 0) return null
  return round(musclePerim ** 2 / (4 * Math.PI), 2)
}

// ── Índice Cintura/Talla ──────────────────────────────────────────────────────

export function calcWaistHeightRatio(
  waist_cm: number | null,
  height_cm: number | null,
): number | null {
  if (!validNum(waist_cm) || !validNum(height_cm) || height_cm! <= 0) return null
  return round(waist_cm! / height_cm!, 3)
}

export function whtrCategory(ratio: number): string {
  if (ratio < 0.5) return 'Saludable'
  if (ratio < 0.6) return 'Sobrepeso'
  return 'Obesidad'
}

// ── IMC ───────────────────────────────────────────────────────────────────────

export function calcBmi(weight_kg: number | null, height_cm: number | null): number | null {
  if (!validNum(weight_kg) || !validNum(height_cm) || height_cm! <= 0) return null
  return round(weight_kg! / (height_cm! / 100) ** 2, 1)
}

export function bmiCategory(bmi: number): string {
  if (bmi < 18.5) return 'Bajo peso'
  if (bmi < 25.0) return 'Normal'
  if (bmi < 30.0) return 'Sobrepeso'
  if (bmi < 35.0) return 'Obesidad I'
  if (bmi < 40.0) return 'Obesidad II'
  return 'Obesidad III'
}

// ── Clasificación % grasa ─────────────────────────────────────────────────────

export function fatCategory(pct: number, sex: string): string {
  if (isMale(sex)) {
    if (pct < 6)  return 'Esencial'
    if (pct < 14) return 'Atleta'
    if (pct < 18) return 'En forma'
    if (pct < 25) return 'Aceptable'
    return 'Obesidad'
  } else {
    if (pct < 14) return 'Esencial'
    if (pct < 21) return 'Atleta'
    if (pct < 25) return 'En forma'
    if (pct < 32) return 'Aceptable'
    return 'Obesidad'
  }
}
